File: my-solutions/1014-k-closest-points-to-origin/solution.py
import heapq

class Solution:
    def kClosest(self, points: List[List[int]], k: int) -> List[List[int]]:
        def distance(point):
            return (point[0]**2 + point[1]**2)**0.5

        distance_point_map = {}
        minheap = []

        for i in range(len(points)):
            dist = distance(points[i])
            if dist in distance_point_map:
                distance_point_map[dist].append(points[i])
            else:
                distance_point_map[dist] = [points[i]]
            heapq.heappush(minheap, dist)

        res = []
        iter = k
        while iter > 0:
            while iter > 0 and len(distance_point_map[minheap[0]]) > 0:
                res.append(distance_point_map[minheap[0]].pop())
                iter -= 1
            heapq.heappop(minheap)
        
        return res

File: my-solutions/1014-k-closest-points-to-origin/solution.cpp
class Solution {
public:
    vector<vector<int>> kClosest(vector<vector<int>>& points, int k) {
        auto dist = [](vector<int> a) {
            return a[0] * a[0]  + a[1] * a[1];
        };
        auto cmp = [dist](vector<int> a, vector<int> b){
            return dist(a) < dist(b);
        };
        priority_queue<vector<int>, vector<vector<int>>, decltype(cmp)> kClosest(cmp);
        vector<vector<int>> res;
        kClosest.push(points[0]);
        for(int i = 1; i < points.size(); ++i) {
            if(dist(points[i]) < dist(kClosest.top()) || kClosest.size() < k) kClosest.push(points[i]);
            if(kClosest.size() > k) kClosest.pop();
        }
        while(!kClosest.empty()) {
            res.push_back(kClosest.top());
            kClosest.pop();
        }
        return res;
    }
};
